### ISSUE TITLE
Deleting stale generated classes is NOT working properly

### ISSUE THREAD

**Issue by gstevey** [gradle/incap#109]
I have a small sample app that uses an incremental annotation processor.

After a clean build:

find . -name "*.class"
./app/build/classes/java/main/App.class
./app/build/classes/java/main/AppIncremental.class
./inc/build/classes/java/main/IncapProcessor.class
./inc/build/classes/java/main/Incremental.class

If I then comment out the `@Incremental` annotation and run the build, it succeeds (with ':app:compileJava - all annotation processors are incremental.' in the output).  But:

find . -name "*.class"
./app/build/classes/java/main/App.class
./app/build/classes/java/main/AppIncremental.class
./inc/build/classes/java/main/IncapProcessor.class
./inc/build/classes/java/main/Incremental.class

The AppIncremental generated class has not been deleted.

So basically M1 is not working properly yet.

**Comment by gstevey** [comment]
Update:  This bug is being caused by App.class having an empty dependents set.  This means that either the mapping file is being written incorrectly, or I'm reading it incorrectly into the dependency graph.  Should hopefully be easy to track down and fix.

**Comment by gstevey** [comment]
@stephanenicolas -- here is the source code for my test incap AP.  What am I doing wrong?

```
import java.util.Set;
import java.util.Collections;
import java.io.Writer;
import javax.lang.model.SourceVersion;
import javax.lang.model.util.Elements;
import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.tools.JavaFileObject;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.tools.Diagnostic;
import org.gradle.incap.BaseIncrementalAnnotationProcessor;
            
public class IncapProcessor extends BaseIncrementalAnnotationProcessor {
    private Elements elementUtils;
    private Filer filer;
    private Messager messager;
    
    @Override
    public Set<String> getSupportedAnnotationTypes() {
        return Collections.singleton(Incremental.class.getName());
    }
            
    @Override
    public SourceVersion getSupportedSourceVersion() {
        return SourceVersion.latestSupported();
    }
    
    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        elementUtils = processingEnv.getElementUtils();
        messager = processingEnv.getMessager();
        super.init(processingEnv);
    }
    
    @Override
    public boolean incrementalProcess(Set<? extends TypeElement> elements, RoundEnvironment roundEnv) {
        filer = incrementalProcessingEnvironment.getFiler();
        for (TypeElement annotation : elements) {
            if (annotation.getQualifiedName().toString().equals(Incremental.class.getName())) {
                for (Element element : roundEnv.getElementsAnnotatedWith(annotation)) {
                    TypeElement typeElement = (TypeElement) element;
                    String helperName = typeElement.getSimpleName().toString() + "Incremental";
                    try {
                        JavaFileObject sourceFile = filer.createSourceFile(helperName, element);
                        Writer writer = sourceFile.openWriter();
                        try {
                            writer.write("class " + helperName + " {");
                            writer.write("    String getValue() { return \"incremental\"; }");
                            writer.write("}");
                        } finally {
                            writer.close();
                        }
                    } catch (Exception e) {
                        messager.printMessage(Diagnostic.Kind.ERROR, "Failed to generate source file " + helperName, element);
                    }
                }
            }
        }
        return true;
    }
}
```

**Comment by gstevey** [comment]
Update:  It looks like the mapping file is being written correctly:

App -- SOURCE:AppIncremental.java

Probably a small bug in my code.

**Comment by gstevey** [comment]
The bug has to do with compiling multiple modules.  Not sure yet what the right fix is.

When Gradle is compiling my test app, first it compiles `:inc:compileJava` to compile the annotation processor.  When it does this, it passes a working directory to Incap for the `inc` library: `src/integ/inc/build/intermediates/annotationProcessing`.

Then it compiles the `:app:compileJava` target.  It once again passes the working directory to Incap, this time for the `app` target:  `src/integ/app/build/intermediates/annotationProcessing/`.

Incap writes the mapping file to the second directory (app).  However, when I go to read the mappings file, Incap's Version1MappingFileReader has cached the working dir from the first call (inc).

In particular, the `IncapBuildClientFactory` manages a Singleton instance of the `buildClient`.  That instances caches the first working directory passed to it, and ignores any new working directory.

I _think_ that each module probably needs its own mapping file working dir, even if it's unused.  So the fix here is to change `IncapBuildClientFactory` not to have a singleton `BuildClient` instance.

Assigning to @stephanenicolas to fix.

**Comment by gstevey** [comment]
This is fixed now.
